File: packages/gbnf-experiment/data/20260909T153339Z_9de9fd66/ported_implementation/gbnf/gbnf.py
import json

from .grammar_graph.graph import Graph
from .grammar_graph.parse_state import ParseState
from .grammar_graph.types import ValidInput
from .grammar_parser.build_rule_stack import build_rule_stack
from .rules_builder.rules_builder import RulesBuilder
from .utils.errors.grammar_parse_error import GrammarParseError


def GBNF(input: object, initial_string: ValidInput = "") -> ParseState:
    grammar = input if isinstance(input, str) else str(input)
    builder = RulesBuilder(grammar)
    rules, symbol_ids = builder.rules, builder.symbol_ids
    if len(rules) == 0:
        raise GrammarParseError(grammar, 0, "No rules were found")
    if not symbol_ids.has("root"):
        available = json.dumps(list(symbol_ids.keys()))
        raise GrammarParseError(
            grammar,
            0,
            f"Grammar does not contain a root symbol. Available symbols are: {available}",
        )
    root_id = symbol_ids.get("root")

    # holes in `rules` mean a referenced rule was never defined, which the
    # RulesBuilder's validation pass has already rejected by this point.
    for rule_id, rule in enumerate(rules):
        if rule is None:
            raise GrammarParseError(grammar, 0, f"Rule {rule_id} was never defined")
    stacked_rules = [build_rule_stack(rule) for rule in rules]
    graph = Graph(grammar, stacked_rules, root_id)
    return ParseState(graph, graph.add(initial_string))

File: proposed-projects/round-trip-experiment/reverse/20260909T133702Z_8305aa82/20260910T075701Z_f285a71f/ported_implementation/src/gbnf.ts
import { Graph } from './grammar-graph/graph.js';
import { ParseState } from './grammar-graph/parse-state.js';
import type { Rule } from './grammar-graph/type-guards.js';
import type { ValidInput } from './grammar-graph/types.js';
import { buildRuleStack } from './grammar-parser/build-rule-stack.js';
import { RulesBuilder } from './rules-builder/rules-builder.js';
import { GrammarParseError } from './utils/errors/grammar-parse-error.js';

export const GBNF = (
  input: string,
  initialString: ValidInput = '',
): ParseState => {
  const grammar = typeof input === 'string' ? input : String(input);
  const { rules, symbolIds } = new RulesBuilder(grammar);
  if (rules.length === 0) {
    throw new GrammarParseError(grammar, 0, 'No rules were found');
  }
  if (!symbolIds.has('root')) {
    throw new GrammarParseError(
      grammar,
      0,
      `Grammar does not contain a root symbol. Available symbols are: ${JSON.stringify(
        symbolIds.keys(),
      )}`,
    );
  }
  const rootId = symbolIds.get('root');

  const stackedRules: (Rule[][] | null)[] = rules.map((rule) =>
    rule !== null ? buildRuleStack(rule) : null,
  );
  const graph = new Graph(grammar, stackedRules, rootId);
  return new ParseState(graph, graph.add(initialString));
};

export default GBNF;
